proposals/p3763.md

```diff
@@ -0,0 +1,637 @@
+Indeed, it's desirable to support a simple syntactic rule for transforming an
+in-class declaration into an out-of-class definition, both for tooling and for
+programmers.
+
+## Proposal
+
+Replace the terminology ("match" and "agree") and corresponding rules here as
+follows:
+
+-   Two named declarations _declare the same entity_ if they have the same scope
+    and the same name.
+-   One declaration _redeclares_ another if they declare the same entity and the
+    second declaration appears after the first, including the case where the
+    first declaration is imported. In this case, the second declaration is said
+    to be a _redeclaration_ of the first.
+-   Two declarations _differ_ if the sequence of tokens in the declaration
+    following the introducer keyword and the optional scope, up to the semicolon
+    or open brace, is different, except for `unused` modifiers on parameters.
+-   The program is invalid if it contains two declarations of the same entity
+    that differ.
+
+```carbon
+class A {
+  fn F(n: i32);
+  fn G(n: i32);
+}
+
+// ❌ Error, parameter name differs.
+fn A.F(m: i32) {}
+
+// ❌ Error, parameter type differs syntactically.
+fn A.G(n: (i32)) {}
+```
+
+## Details
+
+### Where redeclarations can appear -- recap of #3762
+
+Because declarations always declare their entity within the lexically-enclosing
+package, and the package forms part of the scope, declarations for an entity can
+only appear in a single package.
+
+As specified in #3762, entities can be declared in more than one library: Each
+entity has an owning library, and all declarations of that entity outside the
+owning library are declared `extern`. Some
+[special considerations](#extern-declarations) apply to `extern` declarations.
+All declarations of an entity without the `extern` keyword are required to
+appear in the same library.
+
+As a consequence of those rules, excluding `extern` declarations, all
+declarations of an entity are in the same library. This is important: our chosen
+redeclaration rule closely couples the syntactic choices made in all
+declarations of the same entity, and this would be a significant burden to
+refactoring if the scope across which redeclarations could appear were large.
+Requiring an `api` file and its `impl` files to be changed at the same time is,
+in comparison, a minor burden to refactoring.
+
+### One-definition rule
+
+If the entity is declared in the `api` file of a library, it can be redeclared
+in any `impl` file. A definition is then required in exactly one `impl` file.
+
+If the entity is _not_ declared in the `api` file, and is instead introduced as
+non-`extern` in an `impl` file, then it must be defined in that `impl` file.
+Note that this means that such an entity can only be declared in a single `impl`
+file. Entities shared by multiple `impl` files must be defined in the `api`
+file, typically as `private` members of the library.
```

Comment:
Sorry, what I wrote here wasn't what I meant -- the restriction I was going for is: if you declare something in multiple `impl` files, then it needs to be *declared* in the `api` file, not *defined* in the `api` file. Multiple `impl`s can't share something just by declaring it the same way as each other. Instead, if something is first declared in an `impl` file, then it's local to that `impl` file, and if `impl` files want to communicate with each other, they do that via things declared in the `api` file.

Reply:
That's how I understood it, I think. I'm trying to figure out what the rationale is (and this may be something to put in #3762 instead, given I'm trying to hammer out behaviors there).

My understanding from how this affects diagnostics is:
- _with_ this restriction, at compile time:
  - If an `api` file has a declaration, we can only issue an "undefined" diagnostic if we have some way to say "all `impl`s are being compiled at once" (note, I'm sad about this because I think this will be the common case)
  - If an `impl` file has a declaration, we can issue an "undefined" diagnostic.
- _without_ this restriction, at compile time, the behavior is the same either way: no diagnostic unless "all `impl`s are being compiled at once"

In other words, the only diagnostic this enables is for the case it disallows. For that, it takes some incremental flexibility: an API could be fully internal to a library without needing to modify the header.

Reply:
Is this just disallowing the case because you don't think it's helpful?

Reply:
I want redeclarations to be checked during compilation, not at link time, as much as possible; I don't want to have two different `impl` files both contain a declaration that we match up without ever having seen both of them existing at once. And I want to maximize the number of cases where we can diagnose a declaration without a definition during compilation rather than at link time.

C++ lets you play "declare the thing multiple times in independent files that never see each other" games, and it's much more harmful to software evolution than it is helpful. For `extern` declarations, I think we essentially need to put up with it, but I don't think we should be allowing it within the scope of a single library.

Happy to add an "Alternative considered" here if you like.

Reply:
I'm fine with this, but can I just ask to pull this into #3762 so that it's considered alongside the other rules for where declarations may be placed?